### ISSUE TITLE
Search bar for posts

### ISSUE THREAD

**Issue by dag7dev** [tomfran/typo#45]
It would be nice to have a static search bar that doesn't involve an external service to search for text in our articles.

**Comment by tomfran** [comment]
Yes I agree, I briefly looked at some libraries but all of them had custom UI which didn't fit. What do you mean by not using external dependencies?

**Comment by dag7dev** [comment]
Yep, I was exactly thinking about tokenization. With no external libraries, I mean avoid using ElasticSearch or services like that.

**Comment by dag7dev** [comment]
I was thinking about lunr.js, is it feasible for our scopes?

**Comment by tomfran** [comment]
Oh okay, so an external dependency. I think anything would be okay honestly, given that it provides full text search over pages. The main concern would be about the UI.

Ideally, an implementation that allows to manually display search results would be perfect, this way the UI can be tailored to the theme.

**Comment by dag7dev** [comment]
Could be interesting having a separate "seach" page?

**Comment by tomfran** [comment]
Yes we can do that

**Comment by tomfran** [comment]
I am working on an implementation, the idea is to have a Spotlight-like search interface, entirely keyboard based, in the example here you press CMD + / to bring up the search, then use the arrows to navigate and enter to confirm. 

The implementation is based on this: https://gist.github.com/cmod/5410eae147e4318164258742dd05393

https://github.com/user-attachments/assets/50a54b0e-8bba-48a3-ad0e-e1c9a12514fb

Ideally, there will be a /search entry in the menu (configurable maybe) to bring up the search even on mobile.

**Comment by dag7dev** [comment]
That's so nice! Thanks for this!

**Comment by OleMussmann** [comment]
I like the look of it, but I am not a big fan of the user experience.
- Using keyboard shortcuts collides with certain navigation plugins like https://vimium.github.io/ or https://github.com/tridactyl/tridactyl . I'm not sure if screen readers would be affected, too.
- Websites do not use keyboard shortcuts consistently. This means that I would need to remember the specific way of navigation for each website. This is not going to happen.
- It is not discoverable. If you explain the shortcut in some text somewhere, people might miss it. If you put the explanation in a UI element, you could just as well put a search box there.

I'd rather have the search integrate with the existing UI. Maybe something along those lines:

https://github.com/user-attachments/assets/18bb3515-d32d-4ec9-a59f-b38dcf4032a8

I like the selection interface @tomfran  showed, maybe it could attach below the search box?

**Comment by tomfran** [comment]
> * Using keyboard shortcuts collides with certain navigation plugins like https://vimium.github.io/ or https://github.com/tridactyl/tridactyl . I'm not sure if screen readers would be affected, too.

This won't be the only way to access the search bar, a menu entry with `search` will be there, bringing up the search menu. The keyboard shortcuts will be configurable.

> * Websites do not use keyboard shortcuts consistently. This means that I would need to remember the specific way of navigation for each website. This is not going to happen.
> * It is not discoverable. If you explain the shortcut in some text somewhere, people might miss it. If you put the explanation in a UI element, you could just as well put a search box there.

I think you're really thinking too much about this, a menu entry will be more than enough for someone to discover the search bar, and a hint about the keyboard shortcut will be placed somewhere yes.

**Comment by OleMussmann** [comment]
Thanks @tomfran , I understand the idea better now. The way the search results are attached below the search bar in your example video looks really nice. I could imagine those being summoned below the (non-overlay) search bar in the same way, or something similar to it.

Examples of what that would look like are [medium.com](https://medium.com/policy/f03bf92035c9) blogs (fixed search bar) [github.com](https://github.com/tomfran/typo) repos (expanding search bar) or [discourse](https://discourse.nixos.org/t/404-from-cache-nixos-org-preventing-rebuild-on-darwin/59382) forums (overlay search bar, sometimes steals `CTRL+F`).

**Comment by runofthemillgeek** [comment]
@tomfran are you still looking into this?

I'm keen to get this added. I'm also in line with search being a command palette/modal UX but also having a search icon to toggle that somewhere in the header area. 

I was able to get a quick implementation going on using Pagefind and `<dialog>` which was suggested in https://gohugo.io/tools/search/ but it'd mean the user would have to add a separate build step. But otherwise, it's fairly easy.

https://hugomods.com/search/modal/ also looks interesting and doesn't need the build steps. Not sure how easy it is to customize the looks and how good the search functionality is.

**Comment by tomfran** [comment]
@runofthemillgeek feel free to have a look if you want.

> but it'd mean the user would have to add a separate build step. 
Can you elaborate? Since my implementation also required some additional steps.

I would use something pre-made only if it was easy to tweak in appearance, but yeah perhaps it's better to go the custom route.
